bin/faucet/src/handlers.rs

```diff
@@ -55,33 +59,80 @@ pub async fn get_tokens(
 
     // Check that the amount is in the asset amount options
     if !state.config.asset_amount_options.contains(&req.asset_amount) {
-        return Err(ProcessError::BadRequest("Invalid asset amount".to_string()));
+        return Err(HandlerError::BadRequest("Invalid asset amount".to_string()));
     }
 
     let mut client = state.client.lock().await;
 
     // Receive and hex user account id
     let target_account_id = AccountId::from_hex(req.account_id.as_str())
-        .map_err(|err| ProcessError::BadRequest(err.to_string()))?;
-
-    // Execute transaction
-    info!(target: COMPONENT, "Executing mint transaction for account.");
-    let (executed_tx, created_note) = client.execute_mint_transaction(
-        target_account_id,
-        req.is_private_note,
-        req.asset_amount,
-    )?;
-
-    // Run transaction prover & send transaction to node
-    info!(target: COMPONENT, "Proving and submitting transaction.");
-    let block_height = client.prove_and_submit_transaction(executed_tx).await?;
+        .map_err(|err| HandlerError::BadRequest(err.to_string()))?;
+
+    let (created_note, block_height) = loop {
```

Comment:
The loop scares me a bit. We don't actually expect the state to desync -- and something like this happening would constitute a bug, or having spawned multiple faucets with the same configuration and secret keys -- which itself is an issue.

I suggest we do something like:

On faucet startup, fetch the account state. If the account is not yet on chain, then submit a transaction to declare it. This means that the non-startup txns should always have an account seed of `None` (iiuc). In other words, for all `get_tokens` calls the account is known to be on-chain and public etc since these conditions were ensured at startup. This fulfills the requirement to be restartable by itself already.

We then need to decide what to do if `get_tokens` fails due to state mismatch. imo we should crash the faucet 😅 This is really an unexpected condition and means something is really off. However we can decide to at least make this somewhat robust against this condition (though I still argue we don't need to). If we do want then I would rather do something like:

```rust
match client.prove_and_submit_transaction(...).await {
    Err(err) if err.is_state_mismatch() => {
        // update account state and resubmit txn
    }, 
    other => other
}
```
which would only retry once. The current `loop` achieves the same thing ofc, but it does invite trouble if we have a bug elsewhere in the logic causing an infinite retry loop. I would also suggest factoring out the state update code, and the txn proving & submission code to make this easy to read.

Reply:
Yes, it makes sense, thank you. I will rewrite my solution to get initial state from the node on startup and fail if the state was changed unexpectedly.

Reply:
I think I'm OK with the "retry once" strategy in case of state mismatch. Overall, here is my current understanding of what we are doing (correct me if I missed anything):

1. For now we are dealing with public faucets only (we can add support for private faucets later if needed).
2. On startup, we assume that our faucet is already on-chain and try to fetch its most recent state. If that fails, we crush the faucet.
3. During minting, we update the local state and submit the transaction to the node. If the transaction fails, we try to get the latest state from the node and try to submit the transaction again. If anything fails here we crush.

One question is how are the faucets created? For our faucet we can use the faucet created in the genesis block as discussed in https://github.com/0xPolygonMiden/miden-node/pull/517#issuecomment-2430532332. But what about other faucets (this could be an improvement we add later on)?

Reply:
> One question is how are the faucets created?

This actually creates the account if its not there as part of `get_tokens` afaict. Or at least I think that's what the data store's transaction inputs is doing?

Though I'm finding it difficult to keep track 😓 Not this PRs fault; things are quite muddled in the faucet code.

Reply:
I think our end goal should be something like:

Move faucet account creation to a separate subcommand. This should create a random keypair and create the account on chain in `miden-faucet init` (but maybe a better name e.g. `miden-faucet create`).

`miden-faucet start` then loads account state on startup (crashes if missing), and just updates state in `get_tokens`. I don't think there is much point in retrying if the state is different - that would constitute a pretty bad internal failure.

Reply:
> I'm fine with not doing it now if it complicates things, but I think retrying once could be useful. For example, what if there are two parallel instances of the faucet running and they make updates independently to the same account. Unlikely to happen, but also not sure if this should be a "catastrophic failure".

My counter argument would be that exactly one of the two would crash which to me is optimal solution 😬 And we would know about it instead of both just chugging along.

Reply:
I like the solution proposed by @Mirko-von-Leipzig (and developed further by @bobbinth): explicitly create faucet account by special command or import of existing faucet by using account ID and secret. And I don't think that we really need to support two faucet web apps working in parallel, it's better to fail one of them (this significantly simplifies `get_tokens` method). This solution is close to what I'm implementing now, but not exactly the same, so I'm going to update my solution.

Thank you for help!